File: src/core/content/cyclesCity.ts
/**
 * CITY CHAOS'S LADDER - eight cycles of machines, and every boss promoted into the next cycle's
 * elite.
 *
 * ---------------------------------------------------------------------------------------------
 * SEEDED FROM MOSSY'S MEASURED CURVE, AND COMPLETELY INDEPENDENT OF IT
 * ---------------------------------------------------------------------------------------------
 * The HP, speed, damage and XP columns are COPIES of Mossy Mayhem's - which were themselves
 * copies of the Scrapyard's measured pacing - reassigned to fit this map's fiction. Copies and
 * not references, by the same instruction both earlier ladders record: retuning a City cycle
 * must not be able to reach the moss or the yard, and the guarantee is that there is no shared
 * symbol to retune. The three ladders will drift; that is the intended end state.
 *
 * ---------------------------------------------------------------------------------------------
 * THE CASCADE, MECHANICALLY
 * ---------------------------------------------------------------------------------------------
 * A row here names its `regular` and its `boss`; it does NOT name an elite. The elite of cycle N
 * IS `boss` of row N-1 (at the current cycle's body class - see creaturesCity.ts for the two
 * places the class steps up and the returning boss needs a second, bigger row). Row 0 is the one
 * exception and carries the ladder's only authored `elite`.
 *
 * Writing the rule as derivation rather than as eight hand-copied `elite` fields is the same
 * move achievements.ts makes with unlock conditions: the two places that must agree - "who was
 * cycle 3's boss" and "who is cycle 4's elite" - are one field read twice, and cannot drift.
 *
 * ---------------------------------------------------------------------------------------------
 * THE SHAPE OF THE EIGHT
 * ---------------------------------------------------------------------------------------------
 * Only HP and XP climb every rung; speed and contact damage move up AND down, so each cycle asks
 * a different question instead of asking the last one's louder.
 *
 *      1 Junkbots     Slow, soft, everywhere. One enemy for two minutes: learn the streets.
 *      2 Sentries     FAST AND FLIMSY. Unarmed security bots that dart.
 *      3 Drones       THE FASTEST THING ON THE MAP. The air fills up.
 *      4 Rovers       Slow and fat. Six wheels, no hurry, no mercy.
 *      5 Gun Sentries The sentries again, ARMED. Brakes on, damage nearly doubled -
 *                     the first cycle that genuinely hurts to touch.
 *      6 Gun Drones   The drones again, armed. Quick AGAIN, and the hardest hit so far.
 *      7 Fighters     Strafing aircraft. Tanky and unhurried.
 *      8 Armour       A WALL of dozer-yellow tanks. Slowest, heaviest, the end of the ladder.
 *
 * Invariant K: the fastest thing this ladder can produce is a plain Drone regular at 71 u/s,
 * 75.26 with the within-cycle ramp - the same worst case Mossy measured, against the same
 * slowest mech. tests/levels.test.ts re-checks it against this ladder rather than trusting the
 * copy stayed faithful.
 */

import {
  RANK_BOSS,
  RANK_ELITE,
  RANK_REGULAR,
  type ResolvedCycle,
} from './cycles.js';
import { CITY } from './creaturesCity.js';
import { ARCH_BRUISER, ARCH_GRUNT, ARCH_RUNT, type Archetype } from './enemyCatalog.js';

export interface CityCycleDef {
  readonly name: string;
  /** Body class. AUTHORED - must agree with every named creature's drawSize. */
  readonly archetype: Archetype;
  /** The horde. */
  readonly regular: number;
  /**
   * WHO THE RETURNED BOSS IS WHEN IT WALKS THIS CYCLE AS AN ELITE - normally the previous row's
   * `boss`, except where the body class stepped up and the design needs its bigger row (see
   * creaturesCity.ts). Row 0's is the ladder's only authored elite.
   *
   * Stored rather than computed from `boss` alone because of exactly those two seams: the FIELD
   * is the cascade, and the header's promise is kept by the test below the table, which walks
   * every pair of adjacent rows and asserts elite N and boss N-1 draw the same sprite.
   */
  readonly elite: number;
  readonly boss: number;
  /** Regular HP at cycle START, before the within-cycle ramp. */
  readonly hp: number;
  readonly speed: number;
  readonly contactDamage: number;
  readonly xp: number;
  /** P(a regular rolls a non-plain flavour). Zero in cycle 0: the first minute is ONE enemy. */
  readonly variantChance: number;
}

export const CITY_LADDER: readonly CityCycleDef[] = Object.freeze([
  Object.freeze({
    name: 'Junkbots', archetype: ARCH_RUNT as Archetype,
    regular: CITY.ROBOT, elite: CITY.CYB_LARGE, boss: CITY.GEORGE,
    hp: 22, speed: 56, contactDamage: 5, xp: 1, variantChance: 0,
  }),
  Object.freeze({
    name: 'Sentries', archetype: ARCH_RUNT as Archetype,
    regular: CITY.TWOLEGS, elite: CITY.GEORGE, boss: CITY.LEELA,
    hp: 34, speed: 68, contactDamage: 6, xp: 2, variantChance: 0.1,
  }),
  Object.freeze({
    name: 'Drones', archetype: ARCH_RUNT as Archetype,
    regular: CITY.FLYING, elite: CITY.LEELA, boss: CITY.STAN,
    hp: 56, speed: 71, contactDamage: 8, xp: 3, variantChance: 0.16,
  }),
  Object.freeze({
    name: 'Rovers', archetype: ARCH_GRUNT as Archetype,
    regular: CITY.ROVER, elite: CITY.STAN_HEAVY, boss: CITY.MIKE,
    hp: 66, speed: 54, contactDamage: 9, xp: 4, variantChance: 0.22,
  }),
  Object.freeze({
    name: 'Gun Sentries', archetype: ARCH_GRUNT as Archetype,
    regular: CITY.TWOLEGS_GUN, elite: CITY.MIKE, boss: CITY.BEE,
    hp: 104, speed: 53, contactDamage: 14, xp: 6, variantChance: 0.26,
  }),
  Object.freeze({
    name: 'Gun Drones', archetype: ARCH_GRUNT as Archetype,
    regular: CITY.FLYING_GUN, elite: CITY.BEE, boss: CITY.FLAMINGO,
    hp: 118, speed: 65, contactDamage: 18, xp: 8, variantChance: 0.3,
  }),
  Object.freeze({
    name: 'Fighters', archetype: ARCH_GRUNT as Archetype,
    regular: CITY.FIGHTER, elite: CITY.FLAMINGO, boss: CITY.FROG,
    hp: 172, speed: 57, contactDamage: 15, xp: 11, variantChance: 0.32,
  }),
  Object.freeze({
    name: 'Armour', archetype: ARCH_BRUISER as Archetype,
    regular: CITY.TANK, elite: CITY.FROG_HEAVY, boss: CITY.PANDA,
    hp: 225, speed: 50, contactDamage: 22, xp: 15, variantChance: 0.34,
  }),
] as const) as readonly CityCycleDef[];

/** Per-cycle multipliers past the authored ladder. City's own copies, same figures as launch. */
const EXTRA_HP_MUL = 1.45;
const EXTRA_XP_MUL = 1.4;
const EXTRA_DMG_MUL = 1.2;

/**
 * Fills `out` with cycle `index`'s machines.
 *
 * Past the authored ladder it repeats the LAST cycle with compounding multipliers - the honest
 * answer here, as on the moss: minute 17 is more Armour, harder, with the Panda mech still on
 * the boss slot. The compounding is a loop of exact multiplies because `Math.pow` is banned in
 * core (implementation-defined in the last ulp, which is a divergent replay).
 */
export function resolveCityCycle(index: number, out: ResolvedCycle): void {
  const n = CITY_LADDER.length;
  const i = index < n ? index : n - 1;
  const extra = index < n ? 0 : index - (n - 1);
  const def = CITY_LADDER[i];

  let hp = def.hp;
  let xp = def.xp;
  let dmg = def.contactDamage;
  for (let k = 0; k < extra; k++) {
    hp *= EXTRA_HP_MUL;
    xp *= EXTRA_XP_MUL;
    dmg *= EXTRA_DMG_MUL;
  }

  out.index = index;
  out.name = def.name;
  out.archetype = def.archetype;
  out.hp = hp;
  out.speed = def.speed;
  out.contactDamage = dmg;
  out.xp = xp;
  out.variantChance = def.variantChance;

  out.typeByRank[RANK_REGULAR] = def.regular;
  out.typeByRank[RANK_ELITE] = def.elite;
  out.typeByRank[RANK_BOSS] = def.boss;
}

File: cs/src/Scrapyard.Core/CyclesCity.cs
namespace Scrapyard.Core;

/// <summary>
/// Creature ids for City Chaos. Port of <c>CITY</c> in <c>content/creaturesCity.ts</c>. See the
/// remarks on <see cref="MossCreatures"/> - only the ids are ported, never the sprite frames.
/// </summary>
public static class CityCreatures
{
    public const int Robot = 0;
    public const int Twolegs = 1;
    public const int Flying = 2;
    public const int Rover = 3;
    public const int TwolegsGun = 4;
    public const int FlyingGun = 5;
    public const int Fighter = 6;
    public const int Tank = 7;
    public const int CybLarge = 8;
    public const int George = 9;
    public const int Leela = 10;
    public const int Stan = 11;
    public const int StanHeavy = 12;
    public const int Mike = 13;
    public const int Bee = 14;
    public const int Flamingo = 15;
    public const int Frog = 16;
    public const int FrogHeavy = 17;
    public const int Panda = 18;
}

/// <summary>
/// CITY CHAOS'S LADDER. Port of <c>cyclesCity.ts</c>.
/// </summary>
/// <remarks>
/// <b>THE ELITE CASCADE.</b> A rung's elite is normally the PREVIOUS rung's boss walking a size
/// down - George becomes cycle 1's elite, Leela becomes cycle 2's, and so on - except at two seams
/// (<c>StanHeavy</c> at rung 3, <c>FrogHeavy</c> at rung 7) where the body class steps up and the
/// design needs the bigger row instead. Both are AUTHORED per rung rather than computed from
/// <c>Boss</c>, because those two seams exist; a computed cascade would have to special-case them
/// anyway and would then be two places that could disagree about where they are.
/// </remarks>
public static class CityLadder
{
    public readonly struct Rung
    {
        public required string Name { get; init; }
        public required int Archetype { get; init; }
        public required int Regular { get; init; }
        public required int Elite { get; init; }
        public required int Boss { get; init; }
        public required double Hp { get; init; }
        public required double Speed { get; init; }
        public required double ContactDamage { get; init; }
        public required double Xp { get; init; }
        public required double VariantChance { get; init; }
    }

    public static readonly Rung[] All =
    {
        new() { Name = "Junkbots", Archetype = Archetypes.Runt, Regular = CityCreatures.Robot, Elite = CityCreatures.CybLarge, Boss = CityCreatures.George, Hp = 22, Speed = 56, ContactDamage = 5, Xp = 1, VariantChance = 0 },
        new() { Name = "Sentries", Archetype = Archetypes.Runt, Regular = CityCreatures.Twolegs, Elite = CityCreatures.George, Boss = CityCreatures.Leela, Hp = 34, Speed = 68, ContactDamage = 6, Xp = 2, VariantChance = 0.1 },
        new() { Name = "Drones", Archetype = Archetypes.Runt, Regular = CityCreatures.Flying, Elite = CityCreatures.Leela, Boss = CityCreatures.Stan, Hp = 56, Speed = 71, ContactDamage = 8, Xp = 3, VariantChance = 0.16 },
        new() { Name = "Rovers", Archetype = Archetypes.Grunt, Regular = CityCreatures.Rover, Elite = CityCreatures.StanHeavy, Boss = CityCreatures.Mike, Hp = 66, Speed = 54, ContactDamage = 9, Xp = 4, VariantChance = 0.22 },
        new() { Name = "Gun Sentries", Archetype = Archetypes.Grunt, Regular = CityCreatures.TwolegsGun, Elite = CityCreatures.Mike, Boss = CityCreatures.Bee, Hp = 104, Speed = 53, ContactDamage = 14, Xp = 6, VariantChance = 0.26 },
        new() { Name = "Gun Drones", Archetype = Archetypes.Grunt, Regular = CityCreatures.FlyingGun, Elite = CityCreatures.Bee, Boss = CityCreatures.Flamingo, Hp = 118, Speed = 65, ContactDamage = 18, Xp = 8, VariantChance = 0.3 },
        new() { Name = "Fighters", Archetype = Archetypes.Grunt, Regular = CityCreatures.Fighter, Elite = CityCreatures.Flamingo, Boss = CityCreatures.Frog, Hp = 172, Speed = 57, ContactDamage = 15, Xp = 11, VariantChance = 0.32 },
        new() { Name = "Armour", Archetype = Archetypes.Bruiser, Regular = CityCreatures.Tank, Elite = CityCreatures.FrogHeavy, Boss = CityCreatures.Panda, Hp = 225, Speed = 50, ContactDamage = 22, Xp = 15, VariantChance = 0.34 },
    };

    /// <summary>City's own copies, same figures as launch - see the remarks on <see cref="MossyLadder"/>.</summary>
    private const double ExtraHpMul = 1.45;
    private const double ExtraXpMul = 1.4;
    private const double ExtraDmgMul = 1.2;

    public static void Resolve(int index, ResolvedCycle outc)
    {
        int n = All.Length;
        int i = index < n ? index : n - 1;
        int extra = index < n ? 0 : index - (n - 1);
        ref readonly var def = ref All[i];

        double hp = def.Hp;
        double xp = def.Xp;
        double dmg = def.ContactDamage;
        for (int k = 0; k < extra; k++)
        {
            hp *= ExtraHpMul;
            xp *= ExtraXpMul;
            dmg *= ExtraDmgMul;
        }

        outc.Index = index;
        outc.Name = def.Name;
        outc.Archetype = def.Archetype;
        outc.Hp = hp;
        outc.Speed = def.Speed;
        outc.ContactDamage = dmg;
        outc.Xp = xp;
        outc.VariantChance = def.VariantChance;

        outc.TypeByRank[Ranks.Regular] = def.Regular;
        outc.TypeByRank[Ranks.Elite] = def.Elite;
        outc.TypeByRank[Ranks.Boss] = def.Boss;
    }
}

/// <summary>City Chaos.</summary>
public sealed class CityChaosLevel : ILevel
{
    public int CycleCount => CityLadder.All.Length;

    /// <summary>
    /// None. This map opened with Mossy's flock and swapped it for the fuel drums baked into
    /// <see cref="CityBlocks"/> - already the game's loot prop, already drawn, and unlike an animal
    /// it needs no pool, no upkeep tick and no cull radius.
    /// </summary>
    public int Sheep => 0;

    public void ResolveCycle(int index, ResolvedCycle outc) => CityLadder.Resolve(index, outc);

    public string Id => "city-chaos";

    /// <summary>
    /// NO FENCE. The lattice is the boundary here - a player who walks far enough meets more of it,
    /// never an edge - so every clamp that reads this compares against infinity and is never true.
    /// </summary>
    public double ArenaHalf => double.PositiveInfinity;

    public IScenery MakeScenery(int seed) => new CityBlocks(seed);
}
